# **Data Cleaning - Super Store Dataset**

In [1]:
# Importing required libraries
from IPython.display import display, HTML
import pandas as pd

In [2]:
# I will split the Superstore dataset into four tables based on the following data schema:

display(HTML("""
    <div style="width: 100%; text-align: center;">
        <img src="erd_diagram.png" style="width: 100%; max-width: 1150px; height: auto;">
    </div>
"""))

## *Loading and Exploring the Data*

In [3]:
# Loading the dataset using 'latin1' encoding to avoid UnicodeDecodeError
df = pd.read_csv("Superstore.csv", encoding='latin1')

In [4]:
# Displaying dataset information
print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9994 entries, 0 to 9993
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Row ID         9994 non-null   int64  
 1   Order ID       9994 non-null   object 
 2   Order Date     9994 non-null   object 
 3   Ship Date      9994 non-null   object 
 4   Ship Mode      9994 non-null   object 
 5   Customer ID    9994 non-null   object 
 6   Customer Name  9994 non-null   object 
 7   Segment        9994 non-null   object 
 8   Country        9994 non-null   object 
 9   City           9994 non-null   object 
 10  State          9994 non-null   object 
 11  Postal Code    9994 non-null   int64  
 12  Region         9994 non-null   object 
 13  Product ID     9994 non-null   object 
 14  Category       9994 non-null   object 
 15  Sub-Category   9994 non-null   object 
 16  Product Name   9994 non-null   object 
 17  Sales          9994 non-null   float64
 18  Quantity

In [5]:
# Previewing first few and last few rows of the dataset
df.head()

Row ID        Order ID  Order Date   Ship Date       Ship Mode Customer ID  \
0       1  CA-2016-152156   11/8/2016  11/11/2016    Second Class    CG-12520   
1       2  CA-2016-152156   11/8/2016  11/11/2016    Second Class    CG-12520   
2       3  CA-2016-138688   6/12/2016   6/16/2016    Second Class    DV-13045   
3       4  US-2015-108966  10/11/2015  10/18/2015  Standard Class    SO-20335   
4       5  US-2015-108966  10/11/2015  10/18/2015  Standard Class    SO-20335   

     Customer Name    Segment        Country             City  ...  \
0      Claire Gute   Consumer  United States        Henderson  ...   
1      Claire Gute   Consumer  United States        Henderson  ...   
2  Darrin Van Huff  Corporate  United States      Los Angeles  ...   
3   Sean O'Donnell   Consumer  United States  Fort Lauderdale  ...   
4   Sean O'Donnell   Consumer  United States  Fort Lauderdale  ...   

  Postal Code  Region       Product ID         Category Sub-Category  \
0       42420   South  FUR-BO-10001798        Furniture    Bookcases   
1       42420   South  FUR-CH-10000454        Furniture       Chairs   
2       90036    West  OFF-LA-10000240  Office Supplies       Labels   
3       33311   South  FUR-TA-10000577        Furniture       Tables   
4       33311   South  OFF-ST-10000760  Office Supplies      Storage   

                                        Product Name     Sales  Quantity  \
0                  Bush Somerset Collection Bookcase  261.9600         2   
1  Hon Deluxe Fabric Upholstered Stacking Chairs,...  731.9400         3   
2  Self-Adhesive Address Labels for Typewriters b...   14.6200         2   
3      Bretford CR4500 Series Slim Rectangular Table  957.5775         5   
4                     Eldon Fold 'N Roll Cart System   22.3680         2   

   Discount    Profit  
0      0.00   41.9136  
1      0.00  219.5820  
2      0.00    6.8714  
3      0.45 -383.0310  
4      0.20    2.5164  

[5 rows x 21 columns]

In [6]:
df.tail()

Row ID        Order ID Order Date  Ship Date       Ship Mode  \
9989    9990  CA-2014-110422  1/21/2014  1/23/2014    Second Class   
9990    9991  CA-2017-121258  2/26/2017   3/3/2017  Standard Class   
9991    9992  CA-2017-121258  2/26/2017   3/3/2017  Standard Class   
9992    9993  CA-2017-121258  2/26/2017   3/3/2017  Standard Class   
9993    9994  CA-2017-119914   5/4/2017   5/9/2017    Second Class   

     Customer ID     Customer Name   Segment        Country         City  ...  \
9989    TB-21400  Tom Boeckenhauer  Consumer  United States        Miami  ...   
9990    DB-13060       Dave Brooks  Consumer  United States   Costa Mesa  ...   
9991    DB-13060       Dave Brooks  Consumer  United States   Costa Mesa  ...   
9992    DB-13060       Dave Brooks  Consumer  United States   Costa Mesa  ...   
9993    CC-12220      Chris Cortes  Consumer  United States  Westminster  ...   

     Postal Code  Region       Product ID         Category Sub-Category  \
9989       33180   South  FUR-FU-10001889        Furniture  Furnishings   
9990       92627    West  FUR-FU-10000747        Furniture  Furnishings   
9991       92627    West  TEC-PH-10003645       Technology       Phones   
9992       92627    West  OFF-PA-10004041  Office Supplies        Paper   
9993       92683    West  OFF-AP-10002684  Office Supplies   Appliances   

                                           Product Name    Sales  Quantity  \
9989                             Ultra Door Pull Handle   25.248         3   
9990  Tenex B1-RE Series Chair Mats for Low Pile Car...   91.960         2   
9991                              Aastra 57i VoIP phone  258.576         2   
9992  It's Hot Message Books with Stickers, 2 3/4" x 5"   29.600         4   
9993  Acco 7-Outlet Masterpiece Power Center, Wihtou...  243.160         2   

      Discount   Profit  
9989       0.2   4.1028  
9990       0.0  15.6332  
9991       0.2  19.3932  
9992       0.0  13.3200  
9993       0.0  72.9480  

[5 rows x 21 columns]

In [7]:
# Checking the shape
df.shape

(9994, 21)

## *Data Cleaning*

### *Step 1: Standardizing Column Names*

In [8]:
# Standardizing column names (removing spaces and dashes, converting to lowercase)
df.columns = df.columns.str.strip().str.lower().str.replace(' ', '_').str.replace('-', '_')

# Displaying updated column names
df.columns

Index(['row_id', 'order_id', 'order_date', 'ship_date', 'ship_mode',
       'customer_id', 'customer_name', 'segment', 'country', 'city', 'state',
       'postal_code', 'region', 'product_id', 'category', 'sub_category',
       'product_name', 'sales', 'quantity', 'discount', 'profit'],
      dtype='object')

### *Step 2: Checking for Missing Values*

In [9]:
# Checking for null
df.isnull().sum()

row_id           0
order_id         0
order_date       0
ship_date        0
ship_mode        0
customer_id      0
customer_name    0
segment          0
country          0
city             0
state            0
postal_code      0
region           0
product_id       0
category         0
sub_category     0
product_name     0
sales            0
quantity         0
discount         0
profit           0
dtype: int64

### *Step 3: Checking for Duplicates*

In [10]:
# Identifying duplicate rows (excluding the first column to avoid unique key conflicts)
duplicate = df[df.iloc[:,1:].duplicated(keep=False)]  

# Displaying duplicate rows
duplicate

row_id        order_id order_date  ship_date       ship_mode  \
3405    3406  US-2014-150119  4/23/2014  4/27/2014  Standard Class   
3406    3407  US-2014-150119  4/23/2014  4/27/2014  Standard Class   

     customer_id   customer_name      segment        country      city  ...  \
3405    LB-16795  Laurel Beltran  Home Office  United States  Columbus  ...   
3406    LB-16795  Laurel Beltran  Home Office  United States  Columbus  ...   

     postal_code  region       product_id   category sub_category  \
3405       43229    East  FUR-CH-10002965  Furniture       Chairs   
3406       43229    East  FUR-CH-10002965  Furniture       Chairs   

                                           product_name    sales  quantity  \
3405  Global Leather Highback Executive Chair with P...  281.372         2   
3406  Global Leather Highback Executive Chair with P...  281.372         2   

      discount   profit  
3405       0.3 -12.0588  
3406       0.3 -12.0588  

[2 rows x 21 columns]

In [11]:
# Removing duplicate rows
df = df.iloc[:,1:].drop_duplicates()

# Checking the new shape after removing duplicates
df.shape

(9993, 20)

### *Step 4: Converting Data Types*

In [12]:
# Checking data types of each column
df.dtypes

order_id          object
order_date        object
ship_date         object
ship_mode         object
customer_id       object
customer_name     object
segment           object
country           object
city              object
state             object
postal_code        int64
region            object
product_id        object
category          object
sub_category      object
product_name      object
sales            float64
quantity           int64
discount         float64
profit           float64
dtype: object

In [13]:
# Converting 'order_date' and 'ship_date' from MM/DD/YYYY format to DateTime format
df['order_date'] = pd.to_datetime(df['order_date'], format= '%m/%d/%Y')
df['ship_date'] = pd.to_datetime(df['ship_date'], format= '%m/%d/%Y')

# Displaying converted date columns
df[['order_date', 'ship_date']].head()

order_date  ship_date
0 2016-11-08 2016-11-11
1 2016-11-08 2016-11-11
2 2016-06-12 2016-06-16
3 2015-10-11 2015-10-18
4 2015-10-11 2015-10-18

In [14]:
# Checking for any instances where 'ship_date' is earlier than 'order_date'
df[df['ship_date'] < df['order_date']]

Empty DataFrame
Columns: [order_id, order_date, ship_date, ship_mode, customer_id, customer_name, segment, country, city, state, postal_code, region, product_id, category, sub_category, product_name, sales, quantity, discount, profit]
Index: []

In [15]:
# Converting 'postal_code' to string/object type
df['postal_code'] = df['postal_code'].astype('object')
print(df['postal_code'].dtype)

object


In [16]:
# Optimizing Data Types Further
df['quantity'] = df['quantity'].astype('int16')  # Since it's usually a small number
df['discount'] = df['discount'].astype('float16')  # To save memory
print(df[['quantity', 'discount']].dtypes)

quantity      int16
discount    float16
dtype: object


### *Step 5: Checking for Unexpected Values in Categorical Columns*

In [17]:
# Checking unique values in categorical columns
print(df['ship_mode'].value_counts())  
print(df['segment'].value_counts())
print(df['country'].value_counts())  
print(df['region'].value_counts())
print(df['category'].value_counts())
print(df['sub_category'].value_counts())

ship_mode
Standard Class    5967
Second Class      1945
First Class       1538
Same Day           543
Name: count, dtype: int64
segment
Consumer       5191
Corporate      3020
Home Office    1782
Name: count, dtype: int64
country
United States    9993
Name: count, dtype: int64
region
West       3203
East       2847
Central    2323
South      1620
Name: count, dtype: int64
category
Office Supplies    6026
Furniture          2120
Technology         1847
Name: count, dtype: int64
sub_category
Binders        1523
Paper          1370
Furnishings     957
Phones          889
Storage         846
Art             796
Accessories     775
Chairs          616
Appliances      466
Labels          364
Tables          319
Envelopes       254
Bookcases       228
Fasteners       217
Supplies        190
Machines        115
Copiers          68
Name: count, dtype: int64


### *Step 6: Checking for Outliers*

In [18]:
# Generating descriptive statistics
df.describe()

order_date                      ship_date  \
count                           9993                           9993   
mean   2016-04-30 01:53:33.089162240  2016-05-04 00:53:19.039327488   
min              2014-01-03 00:00:00            2014-01-07 00:00:00   
25%              2015-05-23 00:00:00            2015-05-27 00:00:00   
50%              2016-06-26 00:00:00            2016-06-29 00:00:00   
75%              2017-05-14 00:00:00            2017-05-18 00:00:00   
max              2017-12-30 00:00:00            2018-01-05 00:00:00   
std                              NaN                            NaN   

              sales     quantity     discount       profit  
count   9993.000000  9993.000000  9993.000000  9993.000000  
mean     229.852846     3.789753     0.156250    28.660971  
min        0.444000     1.000000     0.000000 -6599.978000  
25%       17.280000     2.000000     0.000000     1.731000  
50%       54.480000     3.000000     0.199951     8.671000  
75%      209.940000     5.000000     0.199951    29.364000  
max    22638.480000    14.000000     0.799805  8399.976000  
std      623.276074     2.225149     0.206421   234.271476

Here, minimum profit value is -6599.978000, which is a large loss and may be an outlier. If a product is sold at a high discount, the selling price may drop below the cost price, resulting in a loss on that sale.

In [19]:
# Filtering rows with negative profit
negative_profit = df[df['profit'] < 0]

# Sorting these rows by profit from lowest (most negative) to highest (least negative)
negative_profit_sorted = negative_profit.sort_values(by='profit')

# Displaying key columns to investigate the issue
columns_to_check = ['order_id', 'product_id', 'profit', 'discount', 'quantity', 'sales']
negative_profit_sorted[columns_to_check].head(15)  # Show top 20 records with most negative profit

order_id       product_id     profit  discount  quantity  \
7772  CA-2016-108196  TEC-MA-10000418 -6599.9780  0.700195         5   
683   US-2017-168116  TEC-MA-10004125 -3839.9904  0.500000         4   
9774  CA-2014-169019  OFF-BI-10004995 -3701.8928  0.799805         8   
3011  CA-2017-134845  TEC-MA-10000822 -3399.9800  0.700195         5   
4991  US-2017-122714  OFF-BI-10001120 -2929.4845  0.799805         5   
3151  CA-2015-147830  TEC-MA-10000418 -2639.9912  0.700195         2   
5310  CA-2017-131254  OFF-BI-10003527 -2287.7820  0.799805         6   
9639  CA-2015-116638  FUR-TA-10000198 -1862.3124  0.399902        13   
1199  CA-2016-130946  OFF-BI-10004995 -1850.9464  0.799805         4   
2697  CA-2014-145317  TEC-MA-10002412 -1811.0784  0.500000         6   
27    US-2015-150630  FUR-BO-10004834 -1665.0522  0.500000         7   
3324  CA-2014-165309  OFF-BI-10001359 -1480.0335  0.799805         5   
165   CA-2014-139892  TEC-MA-10000822 -1359.9920  0.399902         8   
2928  US-2017-120390  OFF-BI-10004995 -1306.5504  0.700195         4   
7898  CA-2017-128363  OFF-BI-10001359 -1237.8462  0.700195         6   

          sales  
7772   4499.985  
683    7999.980  
9774   2177.584  
3011   2549.985  
4991   1889.990  
3151   1799.994  
5310   1525.188  
9639   4297.644  
1199   1088.792  
2697  22638.480  
27     3083.430  
3324    896.990  
165    8159.952  
2928   1633.188  
7898   1614.582

The output shows that the orders with the most negative profits tend to have high discount values. This can be an indication that the high discount is driving the sales price so low that the profit margin turns negative, perhaps even resulting in a loss when compared to the underlying cost.

## *Spliting Data into Separate Tables & Saving the cleaned data*

In [20]:
# Extracting details for seperate tables
customers = df[['customer_id', 'customer_name', 'segment']].drop_duplicates()
orders = df[['order_id', 'customer_id', 'order_date', 'ship_date', 'ship_mode', 'region', 'country', 'state', 'city', 'postal_code']].drop_duplicates()
products = df.groupby('product_id').agg({
    'product_name': lambda x: ', '.join(set(x)),  # Merges multiple product names with same product id
    'category': 'first',  # Takes the first category (as they are the same)
    'sub_category': 'first'  # Takes the first sub-category
}).reset_index()
sales = df[['order_id', 'product_id', 'quantity', 'sales', 'discount', 'profit']].drop_duplicates()

In [21]:
# Making sure all Primary Keys are Unique
duplicate_customers = customers[customers.duplicated(subset=['customer_id'], keep=False)]
duplicate_orders = orders[orders.duplicated(subset=['order_id'], keep=False)]
duplicate_products = products[products.duplicated(subset=['product_id'], keep=False)],
duplicate_sales = sales[sales.duplicated(subset=['order_id', 'product_id'], keep=False)]

print(duplicate_customers, duplicate_orders, duplicate_products, duplicate_sales)

Empty DataFrame
Columns: [customer_id, customer_name, segment]
Index: [] Empty DataFrame
Columns: [order_id, customer_id, order_date, ship_date, ship_mode, region, country, state, city, postal_code]
Index: [] (Empty DataFrame
Columns: [product_id, product_name, category, sub_category]
Index: [],)             order_id       product_id  quantity     sales  discount    profit
350   CA-2016-129714  OFF-PA-10001970         2    24.560  0.000000   11.5432
352   CA-2016-129714  OFF-PA-10001970         4    49.120  0.000000   23.0864
430   US-2016-123750  TEC-AC-10004659         7   408.744  0.199951   76.6395
431   US-2016-123750  TEC-AC-10004659         5   291.960  0.199951   54.7425
1300  CA-2016-137043  FUR-FU-10003664         6   572.760  0.000000  166.1004
1301  CA-2016-137043  FUR-FU-10003664         3   286.380  0.000000   83.0502
3183  CA-2017-152912  OFF-ST-10003208         9  1633.140  0.000000  473.6106
3184  CA-2017-152912  OFF-ST-10003208         3   544.380  0.000000  157.8702


We see that in the sales table there are duplicate entries with different quantities of good sold. If an order includes the same product multiple times, summing ensures total quantity is accurate.

In [23]:
# Aggregating duplicate (order_id, product_id) entries
sales_cleaned = sales.groupby(['order_id', 'product_id'], as_index=False).agg({
    'quantity': 'sum',         # Sum quantity for correct total
    'sales': 'sum',            # Sum sales
    'discount': 'mean',        # Average discount if it varies
    'profit': 'sum'            # Sum profit
})

# Rounding numeric columns to 2 decimal places
sales_cleaned[['sales', 'discount', 'profit']] = sales_cleaned[['sales', 'discount', 'profit']].round(2)

In [24]:
# Saving CSV files
customers.to_csv("customers.csv", index=False)
orders.to_csv("orders.csv", index=False)
products.to_csv("products.csv", index=False)
sales_cleaned.to_csv("sales.csv", index=False)

Now I will continue my analysis using SQL.